source/clast.c

```diff
@@ -2110,3 +2114,463 @@ void clast_filter(struct clast_stmt *node,
+}
+/* Unroll jams a statement. Current supports unroll jam of a single statement as
+ * opposed to a statement block. */
+void unroll_jam_statement(struct clast_stmt *s, unsigned ufactor,
+        const char *uj_iterator, cloog_int_t stride) {
+    struct clast_user_stmt *in;
+    struct clast_stmt *curr, *prev, *last, *sub, **unroll_stmts;
+    int num_unroll_stmts, i, j;
+
+    assert(CLAST_STMT_IS_A(s, stmt_user));
+    unroll_stmts = get_statements_to_be_unrolled(s, &num_unroll_stmts);
+    last = unroll_stmts[num_unroll_stmts-1];
+    prev = last;
+    struct clast_stmt *out = last->next;
+    for (i = 1; i < ufactor; i++) {
+        /* The new unroll offset is i*loop_stride */
+        cloog_int_t uf;
+        cloog_int_init(uf);
+        cloog_int_set_si(uf, i);
+        /* Updated access is (iterator + (i*stride)). Compute i*stride first. */
+        cloog_int_mul(uf, uf,stride);
+        for (j = 0; j< num_unroll_stmts; j++) {
+            in = (struct clast_user_stmt *)unroll_stmts[j];
+            /* sub = clast_substitutions_dup(in->substitutions); */
+            sub = clast_ast_dup(in->substitutions);
+            update_iterator_in_substitution(sub, uj_iterator, uf);
+            curr = &new_clast_user_stmt(in->domain, in->statement, sub)->stmt;
+            prev->next = curr;
+            prev = curr;
+        }
+        cloog_int_clear(uf);
+    }
+    prev->next = out;
+    free(unroll_stmts);
+}
+
+/* Updates loop stride after unroll jamming. The new stride is given
+ * by unroll (factor * loop stride). */
+void update_loop_stride(struct clast_for *loop) {
+    int ufactor;
+    cloog_int_t uf;
+    ufactor = loop->ufactor;
+    cloog_int_init(uf);
+    cloog_int_set_si(uf, ufactor);
+    cloog_int_mul(loop->stride, loop->stride, uf);
+    cloog_int_clear(uf);
+}
+
+/* Unroll jams a loop body. */
+void unroll_jam_loop_body(struct clast_stmt *s, int ufactor,
+        const char *iterator, cloog_int_t stride) {
+    struct clast_for *loop;
+    struct clast_guard *g;
+    if (!s)
+        return;
+    if (CLAST_STMT_IS_A(s, stmt_user)) {
+        /* struct clast_stmt *next_stmt = s->next; */
+        unroll_jam_statement(s, ufactor, iterator, stride);
+        /* unroll_jam_loop_body(next_stmt, ufactor, iterator); */
+    } else if (CLAST_STMT_IS_A(s, stmt_for)) {
+        loop = (struct clast_for *)s;
+        unroll_jam_loop_body(loop->body, ufactor, iterator, stride);
+        unroll_jam_loop_body(s->next, ufactor, iterator, stride);
+    } else if (CLAST_STMT_IS_A (s, stmt_guard)){
+        g = (struct clast_guard *)s;
+        unroll_jam_loop_body(g->then, ufactor, iterator, stride);
+        unroll_jam_loop_body(s->next, ufactor, iterator, stride);
+    }
+    return;
+}
+
+/* Updates the loop upper bound during unroll jam. The new upper bound is
+ * given by (UB - (unroll factor-1)* loop stride) where UB is the upper bound
+ * of the original loop. */
+void update_loop_upper_bound (struct clast_for *loop) {
```

Comment:
Please rename these functions since these are unroll/unroll-jam specific. For eg:

update_loop_upper_bound_for_unrolling

or pass additional args necessary.

Reply:
Renamed to `update_upper_bound_for_unrolling`.